### PR TITLE
build: add -fno-math-errno to allow inlining of math functions

### PR DESCRIPTION
**Pull request by kasper93**
In some cases, where there is a lot of simple text on the screen, the ass_render_event and quantize_transform were taking an unexpected amount of CPU time, causing frame drops.

It turns out that lrint is the root cause of this, as it is not inlined. Allow it to be lowered to the cvtsd2si instruction.

This change also helps with other similar cases and allows simple math functions to be inlined.

Overall, rendering performance is 2.2 times faster than before, based on my test case.

### CODE REVIEW

**Review by rcombs — COMMENTED** [review]
Seems entirely sensible to me; just wanna double-check that we're managing our compiler support appropriately.

For the curious: this doesn't affect macOS, since `-fno-math-errno` is set by default there.

**Comment by rcombs** [comment]
Also, could you post the test file you used for the perf benchmarking?

File: Makefile.am [hunk 1]
```diff
@@ -3,7 +3,7 @@ AM_CPPFLAGS = -I$(top_srcdir)/libass
 AM_CFLAGS = -std=gnu99 -Wall -Wextra -Wno-sign-compare -Wno-unused-parameter \
             -Werror-implicit-function-declaration -Wstrict-prototypes        \
             -Wpointer-arith -Wredundant-decls -Wno-missing-field-initializers\
-            -D_GNU_SOURCE
+            -D_GNU_SOURCE -fno-math-errno
```
[the remark this answers is not in the document]
Comment (kasper93) [reply Makefile.am:6]: This list of flags looks already specific to GCC/Clang compilers and both of them support `-fno-math-errno`. For meson the added flags goes through `get_supported_arguments`.

**Comment by kasper93** [comment]
> Also, could you post the test file you used for the perf benchmarking?

I don't have a specific file. I tested with console.lua in mpv, with the same content on the screen. Generally inlining math functions will always be a win. Performance gain will vary depending on number of events and stuff, and also on libc used, but it won't be slower.

**Comment by rcombs** [comment]
Yeah, agreed this is a clear win, I was just hoping you had a test case that hit `lrint` particularly heavily that we could use for benchmarking in the future. Agreed that this is a pretty obvious win regardless.

**Comment by NBruderman** [comment]
> Yeah, agreed this is a clear win, I was just hoping you had a test case that hit `lrint` particularly heavily that we could use for benchmarking in the future. Agreed that this is a pretty obvious win regardless.

@rcombs Do the samples provided here might help?
https://github.com/libass/libass/issues/736

File: Makefile.am (hunk 1 above)
[the remark this answers is not in the document]
Comment (TheOneric) [reply Makefile.am:6]: `-D` is apparently required by a par of SUS, `-std=` is also near universal (not sure if part of a standard)
And other than clang and gcc, most other compilers just ignore unknwon warning options

Passing warning options through config checks to allow toggling some more specific warnings has been desired for a while, but it just didn’t bother anyone enough to actual implement it yet; see #529

`-f` options are more likely to cause issues; CI passing shows tcc either understands or ignores this flag, but e.g. chibicc will fail due to it. Granted, while i compiled libass successfully before with chibicc, right now it fails anyway due to what appears to be a chibicc bug. This notwithstanding i wouldn’t be surprised to find out some currently fully working compiler gets broken by this flag.

File: meson.build [hunk 2]
```diff
@@ -23,7 +23,7 @@ host_system = host_machine.system()
 cc = meson.get_compiler('c')
 
 cc_warnings = []
-cc_features = []
+cc_features = ['-fno-math-errno']
 
 if cc.get_id() != 'msvc'
     cc_features += '-D_POSIX_C_SOURCE=200809L'
```
Comment (TheOneric) [inline meson.build:29]: We know this will never work for MSVC, so we can skip this just like the POSIX define

Comment (kasper93) [reply meson.build]: Ok, moved into the `msvc` branch.

File: Makefile.am (hunk 1 above)
[the remark this answers is not in the document]
Comment (kasper93) [reply Makefile.am]: I've used now `AX_CHECK_COMPILE_FLAG` to check for the flag, if this is a concern. This macro is commonly defined in `autoconf-archive` package, are you ok with that?

[the remark this answers is not in the document]
Comment (rcombs) [reply Makefile.am]: CI's failing on it; I'd expect you'll need to bundle [ax_check_compile_flag.m4](http://git.savannah.gnu.org/gitweb/?p=autoconf-archive.git;a=blob_plain;f=m4/ax_check_compile_flag.m4) in the `m4` directory.

[the remark this answers is not in the document]
Comment (kasper93) [reply Makefile.am]: I've added `autoconf-archive` to ci now, should help, maybe. We can also vendor `ax_check_compile_flag.m4`, I don't know what's preferred here.

[the remark this answers is not in the document]
Comment (rcombs) [reply Makefile.am]: The common behavior with our dependencies (I checked freetype, harfbuzz, fribidi, and fontconfig) is to vendor these.

[the remark this answers is not in the document]
Comment (kasper93) [reply Makefile.am]: Done.

**Review by rcombs — APPROVED** [review]

File: m4/ax_check_compile_flag.m4 [hunk 3]
```diff
@@ -0,0 +1,53 @@
+# ===========================================================================
```
Comment (TheOneric) [inline m4/ax_check_compile_flag.m4:1]: Did you force git to add this file? I think this is currently matched by `.gitignore` rules and should be explicitly excluded

**Review by TheOneric — APPROVED** [review]
Can confirm:
- `ax_check_compile_flag.m4` exactly matches upstream revision 6 `c3896088c00457e02059986bec054bb35dbb23ae`
- we never use `errno` for math functions, only:
    - we read errno only for file and iconv operations
    - our array realloc utils both set and read errno
    - our locale-independent float parser sets errno, but we don't seem to read it

For future extension of checking e.g. warning options `AX_APPEND_COMPILE_FLAGS` is probably more convenient, but for now `AX_CHECK_COMPILE_FLAG` should be fine and the former requires the latter anyway.
We should also really start to document our compiler/language assumptions and requirements somewhere now that standard features are explicitly forbidden to use, but that’s out of scope for this PR.

Apart from the gitignore nit, seems fine

File: m4/ax_check_compile_flag.m4 (hunk 3 above)
Comment (kasper93) [reply m4/ax_check_compile_flag.m4:1]: Done.

**Review by rcombs — APPROVED** [review]

**Comment by kasper93** [comment]
> For future extension of checking e.g. warning options `AX_APPEND_COMPILE_FLAGS` is probably more convenient, but for now `AX_CHECK_COMPILE_FLAG` should be fine and the former requires the latter anyway. We should also really start to document our compiler/language assumptions and requirements somewhere now that also standard features are explicitly forbidden to use, but that’s out of scope for this PR.

I actually like the idea, I've added `AX_APPEND_COMPILE_FLAGS` and its dependencies, if that's ok with you. Will be useful for future use. Adding more flags only needs adding in one array.

**Comment by astiob** [comment]
(Not a blocker:) Should we perhaps use `-ffast-math` while we’re at it? IIRC @MrSmile has been using it for his own builds for years and had no problems. Off the top of my head, my biggest concern would be `ass_strtod` (e. g. it wouldn’t be nice to get it stuck in an infinite loop like those old Java/PHP parsers on x87), but perhaps it’s dumb enough that it wouldn’t be affected?

**Comment by kasper93** [comment]
> Should we perhaps use -ffast-math while we’re at it?

I would advise to avoid `-ffast-math`, especially in default flags. If anyone feels adventurous they can use it, but it is too big of a hammer to use as-is. Well, with careful considerations. For example `isnan` is used in `ass_outline_update_min_transformed_x` and `dtoi32` so `-ffast-math` will break this. `isnan` would be no-op in this case.

> break it down into its individual components

Actually I want to add `-fno-signed-zeros` which is another safe option to enable. (also enabled in ffmpeg by default). But I want to do it in another PR to avoid stalling this one. And it is less important, than errno one, which actually makes a big difference in some cases.

**Review by TheOneric — APPROVED** [review]

**Pull request merged**